### ISSUE TITLE
troubleshooting: fixed heap overhead for cgo invocation

### ISSUE THREAD

**Issue by anuraaga** [tinygo-org/tinygo#3581]
tl;dr Is there any known memory overhead when cgo is part of the AST that manifests at package initialization?

I am looking into heap memory usage in coraza-proxy-wasm. We use nottinygc, so malloc is implemented separately from the Go heap by mimalloc, and the Go heap is implemented by bdwgc.

I understand the slotting of these libraries complicates things compared to vanilla a lot, but wondering if anything intuitive comes to mind just based on symptoms.

The code I am looking at specifically is this cgo call

https://github.com/wasilibs/go-re2/blob/main/internal/cre2/cre2.go#L34

If I stub it out in a constant way, memory usage is low

```
if true {
  return nil
}
```

`Sys: 16646144, HeapSys: 10354688`

Sys is the total allocated memory, the sum of mimalloc (C heap) and bdwgc (Go heap) stats. HeapSys is just the Go heap reported by bdwgc.

If I stub out in a non-constant way, the go heap size blows up

```
if i != 0 {
  return nil
}
i++
```

(`i` is a package variable initialized to 0)

`Sys: 39124992, HeapSys: 32833536`

And without stubbing, e.g. compiling all of about 50 regexes rather than just 1,

`Sys: 43319296, HeapSys: 32833536`

The Go Heap is constant while C heap goes up.

Instrumenting runtime.alloc, I can confirm this codepath does not call it, as expected.

So I'm suspecting that when stubbed out in a constant way, something is removed completely from the AST that is related to the Go heap usage. As this codepath itself is not allocating, I wonder if it is something added to binary initialization itself.

Is there any known memory overhead when cgo is part of the AST that manifests at package initialization?

**Comment by aykevl** [comment]
> Is there any known memory overhead when cgo is part of the AST that manifests at package initialization?

No, but there could be many things that trigger an escape to the heap. I suggest adding `-print-allocs=.` before and after the change to see whether any new allocations were added in your program.
